django_auth_adfs/backend.py

```diff
@@ -322,27 +323,24 @@ def update_user_groups(self, user, claim_groups):
         """
         if settings.GROUPS_CLAIM is not None:
             # Update the user's group memberships
-            django_groups = [group.name for group in user.groups.all()]
+            user_group_names = user.groups.all().values_list("name", flat=True)
 
-            if sorted(claim_groups) != sorted(django_groups):
-                existing_groups = list(Group.objects.filter(name__in=claim_groups).iterator())
-                existing_group_names = frozenset(group.name for group in existing_groups)
-                new_groups = []
+            if sorted(claim_groups) != sorted(user_group_names):
+                # Get the list of already existing groups in one query
+                existing_claimed_groups = Group.objects.filter(name__in=claim_groups)
+                existing_claimed_group_names = (
+                    group.name for group in existing_claimed_groups
+                )
+
+                new_claimed_group_names = (name for name in claim_groups if name not in existing_claimed_group_names)
                 if settings.MIRROR_GROUPS:
-                    new_groups = [
+                    new_claimed_groups = [
                         Group.objects.get_or_create(name=name)[0]
-                        for name in claim_groups
-                        if name not in existing_group_names
+                        for name in new_claimed_group_names
                     ]
                 else:
-                    for name in claim_groups:
-                        if name not in existing_group_names:
-                            try:
-                                group = Group.objects.get(name=name)
-                                new_groups.append(group)
-                            except ObjectDoesNotExist:
-                                pass
-                user.groups.set(existing_groups + new_groups)
+                    new_claimed_groups = Group.objects.filter(name__in=new_claimed_group_names)
```

Comment:
When `settings.MIRROR_GROUPS` is false, the only groups that should be set for the user are `existing_claimed_groups`. You can likely move the list comprehension of 335 into the `if settings.MIRROR_GROUPS` branch and define `new_claimed_groups` as an empty list in the else branch.

Reply:
Yes, indeed!
It does't make sense to filter again on `new_claimed_group_names` because the result won't change.
It avoids one more query, great!

I prefer to move `user.groups.set()` in the `if` branches instead of define `new_claimed_groups` as an empty list.